### PR TITLE
[feat]: [PL-40056]: Adds processing time metrics for outbox events

### PR DESCRIPTION
**Pull request by ashishsanodia-harness**
## Describe your changes
Adds processing time metrics for outbox events

Test:
- Verified below metrics being listed in prometheus metrics scrap endpoint 
  - outbox_event_processing_time

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50600)

### CODE REVIEW

**Comment by ashishsanodia-harness** [comment]
trigger uts

**Comment by ashishsanodia-harness** [comment]
trigger messagecheck

File: 955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java [hunk 1]
```diff
@@ -85,10 +93,16 @@ private void pollAndHandleOutboxEvents() {
 
       for (int i = 0; i < outboxEvents.size() && !Thread.currentThread().isInterrupted(); i++) {
         OutboxEvent outbox = outboxEvents.get(i);
+        long outboxEventWaitingTime = System.currentTimeMillis() - outbox.getCreatedAt();
```
Comment (Raghav-Murali) [inline 955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java:96]: Can we move this line after 97 and use startTime instead ?

File: 400-rest/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# This is outbox metrics template. To use this metric, copy this file to the corresponding module's folder
```
Comment (Raghav-Murali) [inline 400-rest/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml:1]: Why do we have another definition here apart from the ng-manager one above ?

File: 955-outbox-sdk/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# This is outbox metrics template. To use this metric, copy this file to the corresponding module's folder
```
Comment (Raghav-Murali) [inline 955-outbox-sdk/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml:1]: Similar to previous comment - why do we need multiple declarations for these metrics ?

File: srm-service/modules/cv-nextgen-service/service/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# This is outbox metrics template. To use this metric, copy this file to the corresponding module's folder
```
Comment (Raghav-Murali) [inline srm-service/modules/cv-nextgen-service/service/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml:1]: If all these services are using the same outbox library or wrapper then can we just declare the metrics in that library and use a label with `serviceName` similar to how we have `accountId` and `requestPath` labels here https://github.com/harness/harness-core/pull/48628/files. That way each service will pass its respective serviceName as a label and we can have a single declaration for these metrics.

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50600)* status: 19 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml` line 5 at r1](https://reviewable.io/reviews/harness/harness-core/50600#-N_bCB-A0yOlUHpB6CSH:-N_bCB-A0yOlUHpB6CSI:b32vuk) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml#L5)):*
> ```YAML
> identifier: outbox_event_metrics
> metricGroup: outbox_event_metrics_group
> metrics:
> ```

I 
___
*[`120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml` line 7 at r1](https://reviewable.io/reviews/harness/harness-core/50600#-N_bBl6eAoNhMvJ5o24n:-N_bBl6eAoNhMvJ5o24o:b-nnbo7e) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml#L7)):*
> ```YAML
> metrics:
>   - metricName: outbox_event_processing_time
>     metricDefinition: Outbox resource processing time
> ```

Outbox event processing time
___
*[`120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml` line 12 at r1](https://reviewable.io/reviews/harness/harness-core/50600#-N_bBnMWAhjDMlp3Huqt:-N_bBnMWAhjDMlp3Huqu:bvroqz8) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml#L12)):*
> ```YAML
>     unit: "ms"
>   - metricName: outbox_event_waiting_time
>     metricDefinition: Outbox resource waiting time
> ```

Outbox event wait time
___
*[`955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java` line 96 at r1](https://reviewable.io/reviews/harness/harness-core/50600#-N_bDlyM-ZQweJUs-t66:-N_bDlyM-ZQweJUs-t67:b79d6m4) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java#L96)):*
> ```Java
>       for (int i = 0; i < outboxEvents.size() && !Thread.currentThread().isInterrupted(); i++) {
>         OutboxEvent outbox = outboxEvents.get(i);
>         long outboxEventWaitingTime = System.currentTimeMillis() - outbox.getCreatedAt();
> ```

@ashishsanodia-harness I am not sure how much value outboxEventWaitingTime is going to add. I think outbox events pending processing should suffice.

**Comment by ashishsanodia-harness** [comment]
*[`955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java` line 96 at r1](https://reviewable.io/reviews/harness/harness-core/50600#gh-1266150983:-N_cv9KgCHLapgXPMYHq:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java#L96)):*
<details><summary><i>Previously, Raghav-Murali wrote…</i></summary><blockquote>

Can we move this line after 97 and use startTime instead ?
</blockquote></details>

Done.

**Comment by ashishsanodia-harness** [comment]
*[`955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java` line 96 at r1](https://reviewable.io/reviews/harness/harness-core/50600#-N_bDlyM-ZQweJUs-t66:-N_cvGMZBSnAFA6ZwZm4:b-565dlp) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java#L96)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

@ashishsanodia-harness I am not sure how much value outboxEventWaitingTime is going to add. I think outbox events pending processing should suffice.
</blockquote></details>

There are already metrics exposed for outbox_blocked_queue_size and outbox_queue_size

**Comment by ashishsanodia-harness** [comment]
*[`120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml` line 7 at r1](https://reviewable.io/reviews/harness/harness-core/50600#-N_bBl6eAoNhMvJ5o24n:-N_cvYrbE2sX5OsgI57Z:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml#L7)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

Outbox event processing time
</blockquote></details>

Done.

**Comment by ashishsanodia-harness** [comment]
*[`120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml` line 12 at r1](https://reviewable.io/reviews/harness/harness-core/50600#-N_bBnMWAhjDMlp3Huqt:-N_cv_AC2cJgPInbESXK:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml#L12)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

Outbox event wait time
</blockquote></details>

Done.

**Comment by ashishsanodia-harness** [comment]
*[`120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml` line 5 at r1](https://reviewable.io/reviews/harness/harness-core/50600#-N_bCB-A0yOlUHpB6CSH:-N_cvaTHD7tN4Y1oAb5p:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/120-ng-manager/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml#L5)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

I 
</blockquote></details>

Done.

**Comment by ashishsanodia-harness** [comment]
*[`400-rest/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml` line 1 at r1](https://reviewable.io/reviews/harness/harness-core/50600#gh-1266151032:-N_cvcmv7mIecnIuag6n:b-886gsx) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/400-rest/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml#L1)):*
<details><summary><i>Previously, Raghav-Murali wrote…</i></summary><blockquote>

Why do we have another definition here apart from the ng-manager one above ? 
</blockquote></details>

Fixed

**Comment by ashishsanodia-harness** [comment]
*[`srm-service/modules/cv-nextgen-service/service/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml` line 1 at r1](https://reviewable.io/reviews/harness/harness-core/50600#gh-1266157446:-N_cvfWb3ELpGul8fVGF:b-886gsx) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/srm-service/modules/cv-nextgen-service/service/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml#L1)):*
<details><summary><i>Previously, Raghav-Murali wrote…</i></summary><blockquote>

If all these services are using the same outbox library or wrapper then can we just declare the metrics in that library and use a label with `serviceName` similar to how we have `accountId` and `requestPath` labels here https://github.com/harness/harness-core/pull/48628/files. That way each service will pass its respective serviceName as a label and we can have a single declaration for these metrics.
</blockquote></details>

Fixed

**Comment by ashishsanodia-harness** [comment]
*[`srm-service/modules/cv-nextgen-service/service/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml` line 1 at r1](https://reviewable.io/reviews/harness/harness-core/50600#gh-1266157446:-N_cvrbrEC7_WGWS5aiA:bnwyrnl) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/srm-service/modules/cv-nextgen-service/service/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml#L1)):*
<details><summary><i>Previously, ashishsanodia-harness (Ashish Sanodia) wrote…</i></summary><blockquote>

Fixed
</blockquote></details>

We want processing time/wait time metrics based on resourceType. I think this should be suffice.

**Comment by ashishsanodia-harness** [comment]
*[`955-outbox-sdk/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml` line 1 at r1](https://reviewable.io/reviews/harness/harness-core/50600#gh-1266153615:-N_cwH36Bu1jvtvmjmjY:bdzkukm) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/955-outbox-sdk/src/main/resources/metrics/metricDefinitions/outbox_event_metrics.yaml#L1)):*
<details><summary><i>Previously, Raghav-Murali wrote…</i></summary><blockquote>

Similar to previous comment - why do we need multiple declarations for these metrics ?
</blockquote></details>

We want processing time/wait time metrics based on resourceType. that's why created separate metrics/metricsgroup. Let me know if that is not correct

**Comment by ashishsanodia-harness** [comment]
trigger uts

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> Reviewed 3 of 19 files at r1, 5 of 16 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50600)* status: 21 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java` line 96 at r1](https://reviewable.io/reviews/harness/harness-core/50600#-N_bDlyM-ZQweJUs-t66:-N_gIKpg9VaGoj_uCRnW:bdss87n) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java#L96)):*
<details><summary><i>Previously, ashishsanodia-harness (Ashish Sanodia) wrote…</i></summary><blockquote>

There are already metrics exposed for outbox_blocked_queue_size and outbox_queue_size
</blockquote></details>

I don't see them exposed. Can u check they are there? Can we have plot a grafana chart now itself.

Also outboxEventWaitingTime is not helpful. Lets skip it.

**Comment by ashishsanodia-harness** [comment]
trigger uts

**Comment by ashishsanodia-harness** [comment]
*[`955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java` line 96 at r1](https://reviewable.io/reviews/harness/harness-core/50600#-N_bDlyM-ZQweJUs-t66:-N_i0JFZ2lYq4HuXjjMn:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/178e63c425ce9fa23a748194071d0db183ac99c9/955-outbox-sdk/src/main/java/io/harness/outbox/OutboxEventPollJob.java#L96)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

I don't see them exposed. Can u check they are there? Can we have plot a grafana chart now itself.

Also outboxEventWaitingTime is not helpful. Lets skip it.
</blockquote></details>

Done.

**Comment by ashishsanodia-harness** [comment]
trigger ut7

**Comment by ashishsanodia-harness** [comment]
trigger u2

**Comment by ashishsanodia-harness** [comment]
trigger ut2

**Comment by ashishsanodia-harness** [comment]
trigger ut7

**Review by Raghav-Murali — APPROVED** [review]

**Review by ankit-harness — APPROVED** [review]
:lgtm:

> Reviewed 2 of 19 files at r1, 3 of 16 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50600#-:-N_qordCEgWQe_kjq9YB:bnfp4nl)* status: 21 file(s) to review. 4. unresolved discussions.

**Pull request merged**